cdap-app-fabric/src/test/java/io/cdap/cdap/internal/tethering/TetheringClientHandlerTest.java

```diff
@@ -476,6 +478,87 @@ public void testNullEndpoint() throws Exception {
     Assert.assertEquals(HttpResponseStatus.BAD_REQUEST.code(), response.getResponseCode());
   }
 
+  @Test
+  public void testServerReturns404ForConnectedPeer() throws Exception {
+    // Server returns 404 before tethering has been accepted or rejected by admin on server side.
+    serverHandler.setResponseStatus(HttpResponseStatus.NOT_FOUND, NOT_FOUND_MESSAGE);
+
+    // Client initiates tethering with the server
+    createTethering(SERVER_INSTANCE, PROJECT, LOCATION, NAMESPACES, REQUEST_TIME, DESCRIPTION);
+
+    // Server accepts tethering
+    serverHandler.setResponseStatus(HttpResponseStatus.OK);
+
+    String serverEndpoint = serverConfig.getConnectionConfig().getURI().toString();
+    // Tethering state should transition to accepted on the client.
+    waitForTetheringStatus(TetheringStatus.ACCEPTED, SERVER_INSTANCE, serverEndpoint,
+                           PROJECT, LOCATION, NAMESPACES, DESCRIPTION, true);
+
+    // Server return NOT_FOUND on the next poll
+    serverHandler.setResponseStatus(HttpResponseStatus.NOT_FOUND, NOT_FOUND_MESSAGE);
+
+    // Tethering state should transition to rejected on the client.
+    waitForTetheringStatus(TetheringStatus.REJECTED, SERVER_INSTANCE, serverEndpoint,
+                           PROJECT, LOCATION, NAMESPACES, DESCRIPTION, true);
+
+    // cleanup
+    deleteTethering(SERVER_INSTANCE);
+  }
+
+  @Test
+  public void testServerNotFound() throws Exception {
+    Map<String, String> metadata = ImmutableMap.of("project", PROJECT,
+                                                   "location", LOCATION);
+    PeerMetadata peerMetadata = new PeerMetadata(NAMESPACES, metadata, DESCRIPTION);
+    PeerInfo peer = new PeerInfo(SERVER_INSTANCE, serverConfig.getConnectionConfig().getURI().toString(),
+                                 TetheringStatus.ACCEPTED, peerMetadata, REQUEST_TIME);
+
+    // Add peer in ACCEPTED state
+    tetheringStore.addPeer(peer);
+    serverHandler.setResponseStatus(HttpResponseStatus.OK);
+    String serverEndpoint = serverConfig.getConnectionConfig().getURI().toString();
+    waitForTetheringStatus(TetheringStatus.ACCEPTED, SERVER_INSTANCE, serverEndpoint,
+                           PROJECT, LOCATION, NAMESPACES, DESCRIPTION, true);
+
+    // Server is unreachable so 404 error is returned, but not from TetheringServerHandler
+    serverHandler.setResponseStatus(HttpResponseStatus.NOT_FOUND);
+
+    // Tethering status should remain ACCEPTED, but connection becomes inactive
+    Thread.sleep(cConf.getInt(Constants.Tethering.CONNECTION_TIMEOUT_SECONDS) * 1000);
```

Comment:
sleeps in tests are usually a bad idea, they tend to result in flaky tests. It seems like there shouldn't be a need to sleep as the next line tries to wait for the status to be accepted?

Reply:
sleep is not necessary, removed it and fixed the test